flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -1309,6 +1309,25 @@ struct XEmboxOpConversion : public EmboxCommonConversion<fir::cg::XEmboxOp> {
         }
       }
 
+    auto hasSubcomp = !xbox.subcomponent().empty();
+    mlir::Value stepExpr;
+    if (hasSubcomp) {
+      // We have a subcomponent. The step value needs to be the number of
+      // bytes per element (which is a derived type).
+      auto ty0 = base.getType();
+      auto ptrTy = ty0.dyn_cast<mlir::LLVM::LLVMPointerType>();
+      assert(ptrTy && "expected pointer type");
+      auto eleTy = ptrTy.getElementType();
+      while (auto arrTy = eleTy.dyn_cast<mlir::LLVM::LLVMArrayType>())
+        eleTy = arrTy.getElementType();
```

Comment:
It may be easier to work with the FIR types (like the chunk of code above) here, and then convert the FIR element type. One can get the element type of a SequenceType without a loop, for example.